### PR TITLE
Add carbon theme

### LINKED ISSUE #13673: [FEATURE] Add Carbon Theme
### Problem description

I'd like to create prettier dark plots like
https://medium.com/@pYdeas/create-publication-styled-figures-with-hvplot-using-the-bokeh-backend-b623ab48a5d2

### Feature description

Just an additional theme

### Potential alternatives

Use other themes

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by ahuang11**
image

Based on https://medium.com/@pYdeas/create-publication-styled-figures-with-hvplot-using-the-bokeh-backend-b623ab48a5d2

- [x] issues: fixes #13673
- [ ] tests added / passed
- [ ] release document entry (if new feature or API change)

Tested thru holoviews:
```python
import pandas as pd
import hvplot.pandas
import holoviews as hv
from bokeh.themes import Theme

json = {
    "attrs": {
        "figure": {
            "background_fill_color": "#1b1e23",
            "border_fill_color": "#1b1e23",
            "outline_line_alpha": 0,
        },
        "Grid": {
            "grid_line_color": "#808080",
            "grid_line_alpha": 0.1,
        },
        "Axis": {
            # tick color and alpha
            "major_tick_line_color": "#4d4f51",
            "minor_tick_line_alpha": 0,
            # tick labels
            "major_label_text_font": "Courier New",
            "major_label_text_color": "#808080",
            "major_label_text_align": "left",
            "major_label_text_font_size": "0.95em",
            "major_label_text_font_style": "normal",
            # axis labels
            "axis_label_text_font": "Courier New",
            "axis_label_text_font_style": "normal",
            "axis_label_text_font_size": "1.15em",
            "axis_label_text_color": "lightgrey",
            "axis_line_color": "#4d4f51",
        },
        "Legend": {
            "spacing": 8,
            "glyph_width": 15,
            "label_standoff": 8,
            "label_text_color": "#808080",
            "label_text_font": "Courier New",
            "label_text_font_size": "0.95em",
            "label_text_font_style": "bold",
            "border_line_alpha": 0,
            "background_fill_alpha": 0,
            "inactive_fill_alpha": 0,
        },
        "BaseColorBar": {
            # axis labels
            "title_text_color": "lightgrey",
            "title_text_font": "Courier New",
            "title_text_font_size": "0.95em",
            "title_text_font_style": "normal",
            # tick labels
            "major_label_text_color": "#808080",
            "major_label_text_font": "Courier New",
            "major_label_text_font_size": "0.95em",
            "major_label_text_font_style": "normal",
            "background_fill_color": "#1b1e23",
            "major_tick_line_alpha": 0,
            "bar_line_alpha": 0,
        },
        "Title": {
            "text_font": "Courier New",
            "text_font_style": "normal",
            "text_color": "lightgrey",
        },
    }
}

theme = Theme(json=json)

hv.renderer("bokeh").theme = theme

# load in data and postprocess input
df = (
    pd.read_csv(
        "https://mesonet.agron.iastate.edu/cgi-bin/request/daily.py?network=WA_ASOS&stations=SEA&year1=1928&month1=1&day1=1&year2=2023&month2=12&day2=31&var=max_temp_f&var=min_temp_f&var=precip_in&na=blank&format=csv",
        parse_dates=True,
        index_col="day",
    )
    .drop(columns=["station"])
    .astype("float16")
    .assign(
        dayofyear=lambda df: df.index.dayofyear,
        year=lambda df: df.index.year,
    )
)

# get average and 2023
df_avg = df.loc[df["year"].between(1990, 2020)].groupby("dayofyear").mean()
df_2023 = df[df.year == 2023]

# preprocess below/above
df_above = df_2023[["dayofyear", "max_temp_f"]].merge(
    df_avg.reset_index()[["dayofyear", "max_temp_f"]],
    on="dayofyear",
    suffixes=("_2023", "_avg"),
)
df_above["max_temp_f"] = df_above["max_temp_f_avg"]
df_above["max_temp_f"] = df_above.loc[
    df_above["max_temp_f_2023"] >= df_above["max_temp_f_avg"], "max_temp_f_2023"
]

df_below = df_2023[["dayofyear", "max_temp_f"]].merge(
    df_avg.reset_index()[["dayofyear", "max_temp_f"]],
    on="dayofyear",
    suffixes=("_2023", "_avg"),
)
df_below["max_temp_f"] = df_below["max_temp_f_avg"]
df_below["max_temp_f"] = df_below.loc[
    df_below["max_temp_f_2023"] < df_below["max_temp_f_avg"], "max_temp_f_2023"
]

days_above = df_above.query("max_temp_f_2023 >= max_temp_f_avg")["max_temp_f"].size
days_below = df_below.query("max_temp_f_2023 < max_temp_f_avg")["max_temp_f"].size

# create plot elements
plot = df.hvplot(
    x="dayofyear",
    y="max_temp_f",
    by="year",
    color="grey",
    alpha=0.02,
    legend=False,
    hover=False,
)
plot_avg = df_avg.hvplot(x="dayofyear", y="max_temp_f", color="grey", legend=False)
plot_2023 = df_2023.hvplot(x="dayofyear", y="max_temp_f", color="black", legend=False)
scatter_2023 = df_2023.hvplot.scatter(
    x="dayofyear",
    y="max_temp_f",
    color="max_temp_f",
    size=5,
    legend=False,
    cmap="RdBu_r",
    alpha=0.25,
    clabel="TEMPERATURE °F",
)

dark_red = "#FF5555"
dark_blue = "#5588FF"

plot_above = df_above.hvplot.area(
    x="dayofyear", y="max_temp_f_avg", y2="max_temp_f", label="above", legend="top_right"
).opts(fill_alpha=0.2, line_alpha=0.8, line_color=dark_red, fill_color=dark_red)
plot_below = df_below.hvplot.area(
    x="dayofyear", y="max_temp_f_avg", y2="max_temp_f", label="below", legend="top_right"
).opts(fill_alpha=0.2, line_alpha=0.8, line_color=dark_blue, fill_color=dark_blue)

text_days_above = hv.Text(
    35, df_2023["max_temp_f"].max(), f"{days_above}", fontsize=14
).opts(text_align="right", text_baseline="bottom", text_color=dark_red, text_alpha=0.8)
text_days_below = hv.Text(
    35, df_2023["max_temp_f"].max(), f"{days_below}", fontsize=14
).opts(text_align="right", text_baseline="top", text_color=dark_blue, text_alpha=0.8)
text_above = hv.Text(38, df_2023["max_temp_f"].max(), "DAYS ABOVE", fontsize=7).opts(
    text_align="left", text_baseline="bottom", text_color="lightgrey", text_alpha=0.8
)
text_below = hv.Text(38, df_2023["max_temp_f"].max(), "DAYS BELOW", fontsize=7).opts(
    text_align="left", text_baseline="above", text_color="lightgrey", text_alpha=0.8
)

# overlay every ing and save
final = (
    plot
    * plot_avg
    * plot_above
    * plot_below
    * scatter_2023
    * text_days_above
    * text_days_below
    * text_above
    * text_below
).opts(
    xlabel="TIME OF YEAR",
    ylabel="MAX TEMP °F",
    title="SEATTLE 2023 vs AVERAGE (1990-2020)",
    gridstyle={"ygrid_line_alpha": 0},
    xticks=[
        (1, "JAN"),
        (31, "FEB"),
        (59, "MAR"),
        (90, "APR"),
        (120, "MAY"),
        (151, "JUN"),
        (181, "JUL"),
        (212, "AUG"),
        (243, "SEP"),
        (273, "OCT"),
        (304, "NOV"),
        (334, "DEC"),
    ],
    show_grid=True,
    fontscale=1.18,
)
hvplot.show(final)

```

### CODE REVIEW

**Comment by ahuang11** [comment]
@MarcSkovMadsen 
Doesn't look terrible with violin
image

```python
import xarray as xr
import hvplot.xarray
import holoviews as hv
from bokeh.themes import Theme

json = {
    "attrs": {
        "figure": {
            "background_fill_color": "#1b1e23",
            "border_fill_color": "#1b1e23",
            "outline_line_alpha": 0,
        },
        "Grid": {
            "grid_line_color": "#808080",
            "grid_line_alpha": 0.1,
        },
        "Axis": {
            # tick color and alpha
            "major_tick_line_color": "#4d4f51",
            "minor_tick_line_alpha": 0,
            # tick labels
            "major_label_text_font": "Courier New",
            "major_label_text_color": "#808080",
            "major_label_text_align": "left",
            "major_label_text_font_size": "0.95em",
            "major_label_text_font_style": "normal",
            # axis labels
            "axis_label_text_font": "Courier New",
            "axis_label_text_font_style": "normal",
            "axis_label_text_font_size": "1.15em",
            "axis_label_text_color": "lightgrey",
            "axis_line_color": "#4d4f51",
        },
        "Legend": {
            "spacing": 8,
            "glyph_width": 15,
            "label_standoff": 8,
            "label_text_color": "#808080",
            "label_text_font": "Courier New",
            "label_text_font_size": "0.95em",
            "label_text_font_style": "bold",
            "border_line_alpha": 0,
            "background_fill_alpha": 0,
            "inactive_fill_alpha": 0,
        },
        "BaseColorBar": {
            # axis labels
            "title_text_color": "lightgrey",
            "title_text_font": "Courier New",
            "title_text_font_size": "0.95em",
            "title_text_font_style": "normal",
            # tick labels
            "major_label_text_color": "#808080",
            "major_label_text_font": "Courier New",
            "major_label_text_font_size": "0.95em",
            "major_label_text_font_style": "normal",
            "background_fill_color": "#1b1e23",
            "major_tick_line_alpha": 0,
            "bar_line_alpha": 0,
        },
        "Title": {
            "text_font": "Courier New",
            "text_font_style": "normal",
            "text_color": "lightgrey",
        },
    }
}

theme = Theme(json=json)

hv.renderer("bokeh").theme = theme

ds = xr.tutorial.open_dataset("air_temperature")
ds["month"] = ds.time.dt.month
hvplot.show(ds.hvplot.violin(y="air", by="month"))
```

File: src/bokeh/themes/_carbon.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+            "major_label_text_font_size": "0.95em",
+            "major_label_text_font_style": "normal",
+
+            "background_fill_color": "#1b1e23",
+            "major_tick_line_alpha": 0,
+            "bar_line_alpha": 0
+        },
+        "Title": {
+            "text_font": "Courier New",
+            "text_font_style": "normal",
```
Comment (ahuang11) [inline src/bokeh/themes/_carbon.py:68]:
```suggestion
            "text_font_style": "normal",
            "text_font_size": "1.15em",

```

**Comment by ahuang11** [comment]
Does anything else need to be added for this PR?

**Comment by bryevdv** [comment]
@bokeh/dev any thoughts or comments? Generally speaking we've accepted themes as they come but we are starting to accumulate "dark" themes so perhaps we need to make sure to differntiate them in the docs somehow, at least

**Comment by mattpap** [comment]
```
src/bokeh/themes/_carbon.py:64:32: COM812 [*] Trailing comma missing
src/bokeh/themes/_carbon.py:73:6: COM812 [*] Trailing comma missing
```

**Comment by mattpap** [comment]
It would be really nice to have themes under visual integration tests. In particular under "cross" integration tests, which are two part Python and JS visual integration tests, see e.g. [a test case](https://github.com/bokeh/bokeh/blob/branch-3.4/tests/cross/cases/regressions/issue_13660.py) and [a bokehjs visual test](https://github.com/bokeh/bokeh/blob/57f77ac3f0d523f7cfbfb5ea46cd256adf275b4d/bokehjs/test/integration/cross.ts#L35-L38). Though that would require either a new pure Bokeh example/test case or rewriting the existing one in this PR as pure Bokeh.

**Comment by bryevdv** [comment]
@mattpap I think it's reasonable for us to ask new contributors to include things like that once a system has been put in place (by us) that can be referred to and emulated easily. But I think that ought to be hashed out in a separate purpose-written PR. 

@tcmetzger any thoughts about documenting/differentiating themes (more than just presenting them which is all that happens now) or establishing criteria for inclusion?

File: src/bokeh/themes/_carbon.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+            "title_text_font_style": "normal",
+            # tick labels
+            "major_label_text_color": "#808080",
+            "major_label_text_font": "Courier New",
+            "major_label_text_font_size": "0.95em",
+            "major_label_text_font_style": "normal",
+
+            "background_fill_color": "#1b1e23",
+            "major_tick_line_alpha": 0,
+            "bar_line_alpha": 0
```
Comment (ahuang11) [inline src/bokeh/themes/_carbon.py]:
```suggestion
            "bar_line_alpha": 0,
```

File: src/bokeh/themes/_carbon.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        },
+        "Title": {
+            "text_font": "Courier New",
+            "text_font_style": "normal",
+            "text_font_size": "1.15em",
+
+            "text_color": "lightgrey",
+        },
+    }
```
Comment (ahuang11) [inline src/bokeh/themes/_carbon.py]:
```suggestion
    },
```

**Comment by bryevdv** [comment]
@tcmetzger any thoughts about what guidance we might want to adopt around adding new themes going forward?

**Comment by jbednar** [comment]
My vote is to accept any that seem well put together and are general, i.e. not really narrowly tailored in a way that you wouldn't expect them to be useful.  And then if we get a lot of themes, I'd suggest pulling out a few key ones that we recommend and suggest at the front, then leave all the rest in a grab bag. I like having a big palette of themes to choose from, particularly if I'm trying to find something compatible with an existing website where a plot will be embedded.

**Comment by ahuang11** [comment]
Thoughts on getting this merged?

**Comment by bryevdv** [comment]
I'm going to merge this despite my reservations, but will note to @bokeh/core  that I will object up front to merging any future themes until there is some level of minimal testing in place to assure that themes actually continue to function as expected over time, and ideally until there is some agreement about criteria for adding new themes (or where to add them—unlike @jbednar I don't think the official docs should become a ginormous grab-bag of themes. Maybe we could host a theme free-for-all in a separate repo, though)

**Comment by ahuang11** [comment]
Yes a Bokeh theme hub sounds like a good idea!

**Pull request merged**